### ISSUE TITLE
Raspberry pi docker image armhf

### ISSUE THREAD

**Issue by warrmr** [whittlem/pycryptobot#587]
**Describe the bug**
When attempting to run the container from the image repository using the below settings it fails with the following errors down in the container logs.  This suggests that I am attempting to run a x86/x64 image on an armhf system which is likely as I am using the one from the image repository hosted in this repo. 

I have also attempted to build the container for armhf and failed spectacularly. (see followup post in this report. ) 
```
btcusd    | 2021-11-24T19:38:48.485305374Z standard_init_linux.go:228: exec user process caused: exec format error
btcusd    | 2021-11-24T19:38:54.600144045Z standard_init_linux.go:228: exec user process caused: exec format error
btcusd    | 2021-11-24T19:39:01.195989348Z standard_init_linux.go:228: exec user process caused: exec format error
```

**To Reproduce**
docker-compose.yml
```
version: "3.9"

services:
  btcusd:
    image: ghcr.io/whittlem/pycryptobot/pycryptobot:latest
    container_name: btcusd
    volumes:
      - ./market/binance.key:/app/keys/binance.key:ro
      - ./market/BTCBUSD/config.json:/app/config.json
      - ./market/BTCBUSD/pycryptobot.log:/app/pycryptobot.log
      - ./market/BTCBUSD/graphs:/app/graphs
      - /etc/localtime:/etc/localtime:ro
    environment:
      - PYTHONUNBUFFERED=1
    restart: always
```

config.json
```
{
    "binance": {
        "api_url": "https://api.binance.com",
        "config": {
            "base_currency": "BTC",
            "quote_currency": "BUSD",
            "live": 0,
            "sellatloss": 1,
            "disablelog": 0,
            "autorestart": 0,
            "enableinsufficientfundslogging": 1,
            "logbuysellinjson": 1,
            "graphs": 1,
            "websocket": 0,
            "stats": 0,
            "verbose": 1,
            "consolelog": 1,
            "consoleloglevel": "DEBUG"
        },
        "api_key_file": "/app/keys/binance.key"
    }
}
```
**Expected behavior**
Docker container to run and start trading. 

**Desktop (please complete the following information):**
 - Raspberry pi Buster
 - Docker version 20.10.11, build dea9396

**Comment by warrmr** [comment]
Follow up report .

*Git Tag 
- 4.4.2

docker-compose.yml
```
version: "3.9"

services:
  pycrypto-websvc:
    build:
      context: .
      dockerfile: Dockerfile-websvc
    container_name: pycrypto-websvc
    volumes:
      - ./market/:/app/market
      - /etc/localtime:/etc/localtime:ro
    ports:
      - 5000:5000
    environment:
      - PYTHONUNBUFFERED=1
    restart: always
    deploy:
      restart_policy:
        condition: on-failure

  btcusd:
    build:
      context: .
    container_name: btcusd
    volumes:
      - ./market/binance.key:/app/keys/binance.key:ro
      - ./market/BTCBUSD/config.json:/app/config.json
      - ./market/BTCBUSD/pycryptobot.log:/app/pycryptobot.log
      - ./market/BTCBUSD/graphs:/app/graphs
      - /etc/localtime:/etc/localtime:ro
    environment:
      - PYTHONUNBUFFERED=1
    restart: always
    deploy:
      restart_policy:
        condition: on-failure
```

config.json as above. 

The compile on a raspberry pi takes around 7 hours and fails on installing statsmodel with the below error (truncated). I am going to try and see if I can make a working dockerfile that will build for the raspberry pi. Unfortunately at 7 hours per build this could take a while. 

It even takes a good few hours while using buildx on a PC. 

If I solve it I will be sure to submit a pull request. Would appreciate it if someone already knows the answer :) 

```
docker buildx build --platform linux/armhf -f Dockerfile-websvc -t pycryptobot_websvc .
```

```
    copying scipy/_lib/_uarray/setup.py -> build/lib.linux-armv7l-3.9/scipy/_lib/_uarray
    copying scipy/_lib/_uarray/_backend.py -> build/lib.linux-armv7l-3.9/scipy/_lib/_uarray
    copying scipy/_lib/_uarray/__init__.py -> build/lib.linux-armv7l-3.9/scipy/_lib/_uarray
    running build_clib
    customize UnixCCompiler
    customize UnixCCompiler using build_clib
    building 'mach' library
    error: library mach has Fortran sources but no Fortran compiler found
    ----------------------------------------
    ERROR: Failed building wheel for scipy
  Failed to build scipy
  ERROR: Could not build wheels for scipy, which is required to install pyproject.toml-based projects
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/10/26/0fd61f95667e56fd597ecca715dff3623ed1122b6f895ed2b4dfb54afc04/statsmodels-0.12.2.tar.gz#sha256=8ad7a7ae7cdd929095684118e3b05836c0ccb08b6a01fe984159475d174a1b10 (from https://pypi.org/simple/statsmodels/) (requires-python:>=3.6). Command errored out with exit status 1: /app/bin/python /tmp/pip-standalone-pip-lh2axa7l/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-ovkkf8mg/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- setuptools wheel 'cython>=0.29.14' 'numpy==1.15.4; python_version=='"'"'3.6'"'"'' 'numpy==1.15.4; python_version=='"'"'3.7'"'"'' 'numpy==1.17.5; python_version>='"'"'3.8'"'"'' 'scipy>=1.2' Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement statsmodels==0.12.2 (from versions: 0.4.0, 0.4.1, 0.4.3, 0.5.0rc1, 0.5.0, 0.6.0rc1, 0.6.0rc2, 0.6.0, 0.6.1, 0.8.0rc1, 0.8.0, 0.9.0rc1, 0.9.0, 0.10.0rc2, 0.10.0, 0.10.1, 0.10.2, 0.11.0rc1, 0.11.0rc2, 0.11.0, 0.11.1, 0.12.0rc0, 0.12.0, 0.12.1, 0.12.2, 0.13.0rc0, 0.13.0, 0.13.1)
ERROR: No matching distribution found for statsmodels==0.12.2
The command '/bin/sh -c python -m pip install --no-cache-dir -r requirements.txt' returned a non-zero code: 1
ERROR: Service 'pycrypto-websvc' failed to build : Build failed
```

**Comment by warrmr** [comment]
I have updated the dockerfile to add in the raspberry pi sources and install the dependancies required not auto installed on the RPI. 

I am currently running the trading simulator on my Pi 3b+ and has been running for around 10 hours now without any apparent issue. 

Compiled container can be found here. 
https://hub.docker.com/repository/docker/mattwa/pycryptobot

Updated dockerfile pending me making a pull request. 
https://github.com/warrmr/pycryptobot/blob/docker-for-raspberry-pi-armhf/Dockerfile

docker-compose.yml
```
version: "3.9"

services:
    btcusd:
        image: mattwa/pycryptobot
        # image: ghcr.io/whittlem/pycryptobot/pycryptobot:latest
        container_name: btcusd
        volumes:
            - ./market/binance.key:/app/keys/binance.key:ro
            - ./market/BTCBUSD/config.json:/app/config.json
            - ./market/BTCBUSD/pycryptobot.log:/app/pycryptobot.log
            - ./market/BTCBUSD/graphs:/app/graphs
            - /etc/timezone:/etc/timezone:ro
            - /etc/localtime:/etc/localtime:ro
        environment:
            - PYTHONUNBUFFERED=1
        restart: always
```

Config.json
```
{
    "binance": {
        "api_url": "https://api.binance.com",
        "config": {
            "base_currency": "BTC",
            "quote_currency": "BUSD",
            "live": 0,
            "disablebullonly": 1,
            "disablebuyelderray": 1,
            "sellatloss": 1,
            "websocket": 1,
            "disablelog": 1,
            "autorestart": 1,
            "enableinsufficientfundslogging": 1,
            "logbuysellinjson": 1,
            "consolelog": 1,
            "consoleloglevel": "DEBUG"
        },
        "api_key_file": "/app/keys/binance.key"
    }
}

```

**Comment by warrmr** [comment]
Are there any specific tests I should be running against the docker container to make sure it is fully working and not broken. In order to compile for armhf you need to run the following command so the build runners will need to be updated by the repo owner. 

Also are there any contribution guidelines I need to review before submitting my pull request? 

build command for armhf (raspberry pi) 
```
docker buildx build --platform linux/armhf -f Dockerfile-websvc -t pycryptobot_websvc .
```

**Comment by oreaird** [comment]
Just tested this, currently doesn't work on pi-4s, will look into compiling an alternative version.

This is the result of uname:
`Linux XRMS 5.10.63-v8+ #1459 SMP PREEMPT Wed Oct 6 16:42:49 BST 2021 aarch64 GNU/Linux`

**Comment by warrmr** [comment]
@oreaird try the aarch64 version on my docker hub and let us know if it works. I have added aarch64 to the architecture  list, not able to test on a pi4. 

docker pull mattwa/pycryptobot:aarch64

**Comment by paulbruedgam** [comment]
#603 is not really a good and general solution. The Dockerfile in the main branch now contains dev dependencies in a production ready image. The usage of https://www.piwheels.org/simple as extra index would be good for the usecase raspberry pi but not for other architectures. You should it provide as Dockerfile-rpi and repair the multi stage build to remove the dev dependencies from production. A better and general solution would be to provide a Dockerfile that fits the needs for a multi arch docker build. I currently working on a Dockerfile that allows a multi arch build using multi stages and other benefits of docker buildkit.

```bash
docker buildx build --push --platform linux/arm/v7,linux/arm64/v8,linux/amd64 --tag pycryptobot:slim-bullseye .
```
The bots running with the image based on the following Dockerfile. Sometimes there are some small issues. 

```yaml
# syntax=docker/dockerfile:1
###
# Build Image
###
FROM python:3.9-slim-bullseye AS compile-image

RUN DEBIAN_FRONTEND=noninteractive apt-get update -qq && \
    apt-get upgrade -yqq && \
    apt-get install --no-install-recommends -y build-essential python3-dev \
    gfortran libblas-dev liblapack-dev libjpeg62-turbo-dev && \
    rm -rf /var/lib/apt/lists/*

WORKDIR /app

RUN python3 -m venv /app
# Make sure we use the virtualenv:
ENV PATH="/app/bin:$PATH"

COPY requirements.txt .

RUN --mount=type=cache,mode=0755,target=/root/.cache/pip python3 -m pip install --upgrade pip wheel numpy && \
    python3 -m pip install --upgrade --requirement requirements.txt

COPY . /app

###
# Runable Image
###

FROM python:3.9-slim-bullseye

RUN DEBIAN_FRONTEND=noninteractive apt-get update -qq && \
    apt-get upgrade -yqq && \
    apt-get install --no-install-recommends -y libatlas3-base libjpeg62-turbo && \
    rm -rf /var/lib/apt/lists/* && \
    groupadd -g 999 pycryptobot && \
    useradd -r -u 999 -g pycryptobot pycryptobot && \
    mkdir /app && chown -R pycryptobot:pycryptobot /app

WORKDIR /app

COPY --chown=pycryptobot:pycryptobot --from=compile-image /app /app

USER 999

# Make sure we use the virtualenv:
ENV PATH="/app/bin:$PATH"

# Pass parameters to the container run or mount your config.json into /app/
ENTRYPOINT [ "python3", "-u", "pycryptobot.py" ]
```

@warrmr Are you interested in discussing a good solution?

**Comment by warrmr** [comment]
@paulbruedgam yes please I would be very interested in exploring a better solution to the build dependencies problem. 

I need to make a patch anyway as the build pipeline failed as the buildx command seems to want the switches in a specific order. 

I can make a branch in my fork and work on it or I am happy to join you on your branch and work on it. 

Thanks

**Comment by warrmr** [comment]
@paulbruedgam I have had a quick look over your dockerfile, I will give it a good old test and see how it goes. Just so you are aware the reason why I added  https://www.piwheels.org/simple as an extra source was because the compilation of the pandas dependancies were failing after a couple of hours of compiling.  PiWheels give you an alternative repository containing wheel packages cross compiled for the Raspberry pi (mostly armhf) 

I added it in as an Extra Source in order that it will try the default wheels source and then fallback to the raspberry pi version. 

I have noticed that you have removed that and added a number of additional dependancies, gfortran etc It looks like you have found the required dependancies in order to build pandas from scratch without using pinwheels. Is this correct? 

Also is there a specific reason why you chose user 999 is normal convention not to use user 1000? Its probably a Tomato Tomato situation.

This is the typo that I need to submit a patch for.  I got build buildx and build the wrong way around in the git pipelines.  so it might be worth upgrading the dockerfile with the new steps and fixing the pipeline. 

```yaml
          # Build and Publish container image
          docker buidx build --push \
          --tag $IMAGE_NAME $IMAGE_ID:$VERSION \
          --platform linux/armhf,linux/aarch64,linux/amd64 \
          --file Dockerfile .
```

**Comment by paulbruedgam** [comment]
You are correct. I tried to solve the dependencies for all packages. Most time it is really easy. You only have to go to https://www.debian.org/distrib/packages and use the function "Search the contents of packages". With that you can find out which packages provide the shared libraries you need to build the Python extensions. A good way is to use the packages with the prefix dev in the compile stage.

Use id 999 was really bad mistake. I'm working on a better version that uses best practices. You will find the current work under https://github.com/paulbruedgam/pycryptobot/blob/docker/Dockerfile.

It seems that there is a mistake in this command:
```
          # Build and Publish container image
          docker buidx build --push \
          --tag $IMAGE_NAME $IMAGE_ID:$VERSION \
          --platform linux/armhf,linux/aarch64,linux/amd64 \
          --file Dockerfile .
```
The line with --tag want's only one argument. But there is a space between $IMAGE_NAME and $IMAGE_ID:$VERSION. From my point of view there is no need for $IMAGE_NAME in this line. 
```
          # Build and Publish container image
          docker buidx build --push \
          --tag $IMAGE_ID:$VERSION \
          --platform linux/armhf,linux/aarch64,linux/amd64 \
          --file Dockerfile .
```

**Comment by warrmr** [comment]
I have been playing in my own fork at the moment as I have managed to setup GitHub actions so I can properly test the workflow before submitting a PR. I have updated the workflow to make better use of the docker layer caching in the hopes to speed up subsequent build times. 

https://github.com/warrmr/pycryptobot/blob/main/.github/workflows/container.yml

I will give your dockerfile a go and see how it works on my RPI and see what the build times are like as it is the arm64 version that seems to take about 4 hours to build with the pinwheels source in it.  

RE the current pipeline I am going to refrain from submitting a fix for that until we have this improved dockerfile bashed out then we can do it in one hit.

**Comment by warrmr** [comment]
So this is where I have got with your dockerfile I have had to add in a couple of extra dependancies in the run image for pillow to run. 

https://github.com/warrmr/pycryptobot/blob/docker/Dockerfile

```yaml
# syntax=docker/dockerfile:1
###
# Build Image
###
FROM python:3.9-slim-bullseye AS compile-image

# Update the base image and install requirements
# build-essentials: contains build tools you need to build Python extensions
# python3-dev: contains the header files you need to build Python extensions (needed for ARM)
# gfortran:
# libblas-dev:
# liblapack-dev:
# libjpeg62-turbo-dev: needed for build pillow
# https://pillow.readthedocs.io/en/stable/installation.html#building-on-linux
RUN export DEBIAN_FRONTEND=noninteractive \
    && apt-get update \
    && apt-get upgrade --assume-yes \
    && apt-get install --assume-yes --no-install-recommends \
    build-essential=12.9 \
    gfortran=4:10.2.1-1 \
    libblas-dev=3.9.0-3 \
    libjpeg62-turbo-dev=1:2.0.6-4 \
    liblapack-dev=3.9.0-3 \
    python3-dev=3.9.2-3 \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /app

RUN python3 -m venv /app
# Make sure we use the virtualenv:
ENV PATH="/app/bin:$PATH"

RUN pip config --user set global.extra-index-url https://www.piwheels.org/simple

COPY requirements.txt .

# Use docker buildkit for faster builds and cache pip
# hadolint ignore=DL3013,DL3042
RUN --mount=type=cache,mode=0755,target=/tmp/.buildx-cache/.cache/pip python3 -m pip install --upgrade pip wheel \
    && python3 -m pip install --upgrade --requirement requirements.txt

COPY . /app

###
# Runtime Image
###

FROM python:3.9-slim-bullseye AS runtime-image

ARG REPO=whittlem/pycryptobot

LABEL org.opencontainers.image.source https://github.com/${REPO}

# Update the base image and install requirements
# libatlas3-base:
# libjpeg62-turbo: needed to run pillow
RUN export DEBIAN_FRONTEND=noninteractive \
    && apt-get update \
    && apt-get upgrade --assume-yes \
    && apt-get install --assume-yes --no-install-recommends \
    libatlas3-base=3.10.3-10 \
    libjpeg62-turbo=1:2.0.6-4 libopenjp2-7 libtiff5 libxcb1 \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/* \
    && groupadd -g 1000 pycryptobot \
    && useradd -r -u 1000 -g pycryptobot pycryptobot \
    && mkdir -p /app/.config/matplotlib \
    && chown -R pycryptobot:pycryptobot /app

WORKDIR /app

COPY --chown=pycryptobot:pycryptobot --from=compile-image /app /app

USER pycryptobot

# Make sure we use the virtualenv:
ENV PATH="/app/bin:$PATH"

# Make sure we have a config dir for matplotlib when we not the root user
ENV MPLCONFIGDIR="/app/.config/matplotlib"

# Pass parameters to the container run or mount your config.json into /app/
ENTRYPOINT [ "python3", "-u", "pycryptobot.py" ]
```

I have this running on the Raspberry Pi 3b+ and the .dockerignore file is a little too aggressive as I get

```
btcusd    | 2021-12-27T07:01:31.397828105Z Opening websocket to Binance...
btcusd    | 2021-12-27T07:01:31.401971009Z --------------------------------------------------------------------------------
btcusd    | 2021-12-27T07:01:31.403255119Z |                              Python Crypto Bot                               |
btcusd    | 2021-12-27T07:01:31.404465793Z --------------------------------------------------------------------------------
btcusd    | 2021-12-27T07:01:31.406330683Z FileNotFoundError(2, 'No such file or directory')
```

But if i build the image without the ignore it seems to work ok. 

I am going to have a play with the ignore file now I am working with the new dockerfile.

**Comment by warrmr** [comment]
@paulbruedgam I have made a patchfile to blend in my changes into your fork

**Comment by warrmr** [comment]
https://gist.github.com/warrmr/2e80e5654735d80e52afcf4dbf27ca2f

**Comment by warrmr** [comment]
@paulbruedgam I have been testing the dockerfile you posted above and have made some updates to the ignore files and workflow. For ease of bringing the changes in I have created this patch file that you should be able to apply. 
https://gist.github.com/warrmr/2e80e5654735d80e52afcf4dbf27ca2f

**Comment by paulbruedgam** [comment]
Looks good on first view. Apply of the patch won't work. But no problem. I added the changes manually and merged them with my changes.

I still see one optimization. pip currently using the default cache dir. So the cache is not reusable on different builds. I think the command should look like:
```
RUN --mount=type=cache,mode=0755,target=/tmp/.buildx-cache/.cache/pip python3 -m pip install --upgrade pip wheel \
    && python3 -m pip install --upgrade --requirement requirements.txt --cache-dir /tmp/.buildx-cache/.cache/pip
```

I'm adding hadolint in my ci pipeline.
```
    steps:
      - uses: actions/checkout@v2

      - name: Lint Dockerfile
        uses: hadolint/hadolint-action@v1.6.0
        with:
          dockerfile: Dockerfile
```
I pushed my last changes in my docker branch.

**Comment by paulbruedgam** [comment]
The build cache is persistent. On first run the job was cancelled after 6 hours. Now on the second run I found this:
```
#45 [linux/arm/v6 compile-image 6/7] RUN --mount=type=cache,mode=0755,target=/tmp/.buildx-cache/.cache/pip python3 -m pip install --upgrade pip wheel     && python3 -m pip install --upgrade --requirement requirements.txt --cache-dir /tmp/.buildx-cache/.cache/pip
#45 17.21     Uninstalling pip-21.2.4:
#45 17.41       Successfully uninstalled pip-21.2.4
#45 35.08 Successfully installed pip-21.3.1 wheel-0.37.1
#45 46.24 Collecting urllib3>=1.26.5
#45 46.30   Using cached urllib3-1.26.7-py2.py3-none-any.whl (138 kB)
#45 52.19 Collecting pandas==1.1.4
#45 52.24   Using cached pandas-1.1.4.tar.gz (5.2 MB)
#45 ...
```
Pip using the cached wheels and sources. With each run the build should decrease.

**Comment by warrmr** [comment]
You might need to update the `timeout-minutes` setting in the workflow to stop GitHub killing it after 360 minutes (6 hours) 

I have updated mine to 720 minutes (12 Hours as it takes about 11 hours to build on my laptop) 
https://github.com/warrmr/pycryptobot/blob/9cfb934b9bb6bed1b955745a34b07feb5675ad4f/.github/workflows/container.yml#L26

Also I have added beta as a branch to kick off the build along with main that way we end up with 3 tags (latest, beta and Tased version number eg 5.0.1) 
https://github.com/warrmr/pycryptobot/blob/9cfb934b9bb6bed1b955745a34b07feb5675ad4f/.github/workflows/container.yml#L8

**Comment by paulbruedgam** [comment]
Setting timeout-minutes to 720 will not work. Regarding to https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits there is a hard limit at 6 hours.

> Job execution time - Each job in a workflow can run for up to 6 hours of execution time. If a job reaches this limit, the job is terminated and fails to complete.

I think I have an idea to solve this. I will try it today.

**Comment by warrmr** [comment]
I found out the timeout extension doesn't work :) 

Have you had a successful build? I dont think the cache not found is a problem as the cache only uploads on successful build so if no builds have passed then the cache will be empty/incomplete. 

I think you can force a cache upload on fail but im not sure that is a good idea.

**Comment by paulbruedgam** [comment]
I think I have a good way for build the image via GitHub. When it work we should provide two Dockerfiles (Dockerfile and Dockerfile-github). One for build local and one for Github. I have spilt the Dockerfile in multiple stages and run one job per image stage. So all task in each job have 6 hours time to finish.

**Stage 1 - Base**
- Install needed system packages
- Install and update needed python packages (numpy pip setuptools wheel)
- Split requirements.txt in three parts based on dependencies

**Stage 2 - Compile Stage 1**
- Compile and install all expect matplotlib, pandas, statsmodels

**Stage 3 - Compile Stage 2**
- Compile and install matplotlib and pandas

**Stage 4 - Compile Stage 3**
- Compile and install statsmodels

I hope the pipeline goes through successfully.

**Comment by warrmr** [comment]
ARM Docker container is in master :) 
Multi-step build has been re-instated and will be merged into master in PR #652 
Docker metadata (LABEL) tags added to build 

Closing this issue as I am now happy this is resolved :) 


@paulbruedgam if you wish to open an issue RE: 
`RUN pip config --user set global.extra-index-url https://www.piwheels.org/simple`
